### PR TITLE
storage/rocksdb: Add support for specifying merge operator fn and merge commands

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

Initial work towards enabling https://github.com/MaterializeInc/materialize/issues/26003

This adds support for the "merge" operation in rocksdb (https://github.com/facebook/rocksdb/wiki/Merge-Operator) by allowing a user of `mz-rocksdb-util` to provide an 'associative merge function' and to provide 'merge' updates for a key in addition to 'put' and 'delete' updates to a key.

There are 4 semantic changes done to the rocksdb util:
- `bincode::Options` are now part of `InstanceOptions` since they are necessary for use in the merge function
- `InstanceOptions` optionally accepts a `merge_operator` function that can be treated as an 'associative merge operator'
- the `multi_put` command has been renamed `multi_update`, and rather than using an `Option` to demarcate whether the operation should be a Put or a Delete, a new `KeyUpdate` enum has been introduced that also adds support for the `Merge` operation
- there is a `manual_compaction` command added to allow forcing compaction during unit tests

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/rocksdb/src/lib.rs
Comment (guswynn) [inline src/rocksdb/src/lib.rs:96]: this is :chefkiss:

File: src/rocksdb/src/lib.rs
Comment (guswynn) [inline src/rocksdb/src/lib.rs:164]: does this not set only the full merge fn? You need https://docs.rs/rocksdb/latest/rocksdb/struct.Options.html#method.set_merge_operator to also have a partial one, right?

**Review by guswynn — COMMENTED** [review]
Looks good! a few questions!

File: src/rocksdb/src/lib.rs
Comment (rjobanp) [reply src/rocksdb/src/lib.rs:164]: No the point of the associative merge operator is that you only provide one function and it's used for both full/partial merges since the values you're dealing with are always the same. If the values we passed to `merge` were different than `put`, then you need to implement the full merge operator with separate handlers for each.

See https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity
and specifically:
> AssociativeMergeOperator simply inherits from MergeOperator and provides private default implementations of these methods, while exposing a wrapper function for simplicity.

Comment (rjobanp) [reply src/rocksdb/src/lib.rs:164]: I mention that this should be an ` A merge operator to use for associative merges` above but will leave a comment here to link to the wiki and that quote.

Comment (guswynn) [reply src/rocksdb/src/lib.rs:168]: thanks!

**Comment by rjobanp** [comment]
Okay @guswynn this has been refactored to use a trait-based merge operator function so that I can provide a closure rather than an fn-pointer to `rocksdb_util` from the upsert code

File: src/rocksdb/src/lib.rs
Comment (guswynn) [inline src/rocksdb/src/lib.rs:97]: classic rust!

File: src/rocksdb/src/lib.rs
Comment (guswynn) [inline src/rocksdb/src/lib.rs:181]:
```suggestion
                let operands = ValueIterator {
                    iter: existing.chain(operands.iter()),
                    bincode: &bincode,
                    v: std::marker::PhantomData::<V>,
                };
                merge_fn(key, operands).map(|result| bincode.serialize(&result).unwrap())
```

could simplify the api, but not blocking

File: src/rocksdb/tests/basic.rs
Comment (guswynn) [inline src/rocksdb/tests/basic.rs:179]: merges after a put dont merge the original put in right? or do they?

File: src/rocksdb/tests/basic.rs
Comment (guswynn) [inline src/rocksdb/tests/basic.rs:203]:
```suggestion
    // Validate the Get operation returns the correct sum (which is the final put)
```

File: src/rocksdb/tests/basic.rs
Comment (guswynn) [inline src/rocksdb/tests/basic.rs:250]: can we also have a test that just merges? I don't see an actual assert that we `multi_get` a merged value? Just the puts that overwrite the value?

**Review by guswynn — APPROVED** [review]
looks good, a few questions on tests

File: src/rocksdb/tests/basic.rs
Comment (rjobanp) [reply src/rocksdb/tests/basic.rs:250]: Ah yes good catch! I think I lost this after refactoring the tests to validate that `Put` resets the keys too. I'll rework this to ensure we validate the pure-merge workload too

File: src/rocksdb/tests/basic.rs
Comment (rjobanp) [reply src/rocksdb/tests/basic.rs:179]: I'm pretty sure the `Put` basically resets the key at the point its applied, and then merges after that `Put` will see the last `Put` value in the `existing_value` of the merge operator. So merges from before the `Put` are basically dropped

File: src/rocksdb/tests/basic.rs
Comment (rjobanp) [reply src/rocksdb/tests/basic.rs:203]: I'll fix this test instead -- the final count was meant to be based on merge operations, not the final Put value.

File: src/rocksdb/tests/basic.rs
Comment (guswynn) [reply src/rocksdb/tests/basic.rs:179]: Ah but future merges see the previous puts; nice that we will only go from merges to puts in upsert!

File: src/rocksdb/tests/basic.rs
Comment (guswynn) [reply src/rocksdb/tests/basic.rs:250]: thanks!

And a merge-then-put-overriding workload, which is the upsert case!

**Pull request merged**
